### PR TITLE
 add minreadyseconds concept to static pod operator

### PR DESCRIPTION
**Pull request by deads2k**
// minReadySeconds is the time to wait between the completion of an operand becoming ready (all containers ready)
	// and starting the rollout onto the next node.  This avoids a problem with an external load balancer that looks like
	//  1. for some reason we have two instances, maybe a liveness check blipped on one node. it doesn't matter why
	//  2. we bring down an instance on m0 to start a new revision
	//  3. at this point we have one instance running on m1
	//  4. m0 starts up and goes ready, but the LB ready check just timed out and is waiting for X seconds
	//  5. we bring down an instance on m1 to start the new revision.
	//  6. the LB thinks all backends are down and routes randomly
	//  7. no profit.
	// setting this field to 30s can prevent the kube-apiserver from triggering the above flow on AWS.

/hold for proof

### CODE REVIEW

**Comment by sttts** [comment]
Great find!

**Comment by p0lyn0mial** [comment]
"Anything that can go wrong will go wrong" :)

**Comment by p0lyn0mial** [comment]
>  6. the LB thinks all backends are down and routes randomly

is this documented behavior?

File: pkg/operator/staticpod/controller/installer/installer_controller.go [hunk 1]
```diff
@@ -262,6 +281,41 @@ func nodeToStartRevisionWith(ctx context.Context, getStaticPodState func(ctx con
 	return 0, reason, nil
 }
 
+// determine the amount of time to delay before creating the next installer pod.  We delay to avoid
```
Comment (p0lyn0mial) [inline pkg/operator/staticpod/controller/installer/installer_controller.go:284]: nit: fix the gdoc

File: pkg/operator/staticpod/controller/installer/installer_controller.go [hunk 2]
```diff
@@ -284,6 +338,15 @@ func (c *InstallerController) manageInstallationPods(ctx context.Context, operat
...
+	// determine the amount of time to delay before creating the next installer pod.  We delay to avoid an LB outage (see godoc on minReadySeconds)
+	sleepTime := c.timeToWaitBeforeInstallingNextPod(ctx, operatorStatus.NodeStatuses)
+	if sleepTime > 0 {
+		select {
+		case <-ctx.Done():
+		case <-time.After(sleepTime):
+		}
+	}
+
 	for l := 0; l < len(operatorStatus.NodeStatuses); l++ {
```
Comment (p0lyn0mial) [inline pkg/operator/staticpod/controller/installer/installer_controller.go:350]: the code below ensures that only one node can transition to a new revision at a time, right?

File: pkg/operator/staticpod/controller/installer/installer_controller.go [hunk 3]
```diff
@@ -262,6 +281,41 @@ func nodeToStartRevisionWith(ctx context.Context, getStaticPodState func(ctx con
...
+			continue
+		}
+		for _, podCondition := range pod.Status.Conditions {
+			if podCondition.Type != corev1.PodReady {
+				continue
+			}
+			if podCondition.Status != corev1.ConditionTrue {
+				continue
+			}
+			durationPodHasBeenReady := c.clock.Now().Sub(podCondition.LastTransitionTime.Time)
```
Comment (p0lyn0mial) [inline pkg/operator/staticpod/controller/installer/installer_controller.go:304]: okay, it takes `min` from all ready nodes.

**Comment by deads2k** [comment]
> > 1. the LB thinks all backends are down and routes randomly
> 
> is this documented behavior?

Uh, probably documented?  It either routes to all or routes to none. Both outcomes are bad.

File: pkg/operator/staticpod/controller/installer/installer_controller.go (hunk 2 above)
Comment (deads2k) [reply pkg/operator/staticpod/controller/installer/installer_controller.go:350]: > the code below ensures that only one node can transition to a new revision at a time, right?

I can find a ref if you want it, but yes, we check the previous installer pod status

Comment (p0lyn0mial) [reply pkg/operator/staticpod/controller/installer/installer_controller.go:350]: just wanted to make sure that `durationPodHasBeenReady` counting makes sense. thanks.

**Comment by deads2k** [comment]
proof worked, godoc updated.

/hold cancel

**Pull request merged**